host/registry.go

```diff
@@ -0,0 +1,175 @@
+package host
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/core/net/rhp"
+	"go.sia.tech/core/types"
+)
+
+type (
+	registryValidationError struct {
+		err error
+	}
+)
+
+// Error implements the error interface.
+func (e *registryValidationError) Error() string {
+	return e.err.Error()
+}
+
+type registry struct {
+	hostID types.Hash256
+	store  RegistryStore
+
+	// registry entries must be locked while they are being modified
+	registryMu    sync.Mutex
+	registryLocks map[types.Hash256]*locker
+}
+
+// lockRegistryKey locks the registry key with the provided key preventing
+// updates. The context can be used to interrupt if the registry key lock cannot
+// be acquired quickly.
+func (r *registry) lockRegistryKey(key types.Hash256, timeout time.Duration) error {
```

Comment:
since these are methods on the `registry` type, we can probably drop `Registry` from their name

Reply:
ditto for `registryMu` and `registryLocks`